### PR TITLE
[GAN SVS] Add VISinger2, UHifiGAN, Avocodo

### PR DESCRIPTION
**Pull request by jerryuhoo**
@ftshijt @A-Quarter-Mile 
This is an update for VISinger, I added multiple modules.
- [x] Update VISinger 1 generator
  - [x] Add an option to use phoneme predictor or not
  - [x] Add an option to use flow or not
- [x] Add VISinger 2 generator
- [x] Add VISinger 2 vocoder generator (DDSP)
- [x] Add VISinger 2 vocoder discriminator (+MFD)
- [x] Add Avocodo
- [x] Add UHifiGAN
- [x] Add yin feature (For Pits)
  - [ ] Pits
- [x] Unit tests for all those changes 
- [x] Fix F0 bug for SVS

### CODE REVIEW

File: espnet2/gan_svs/pits/pisinger_generator.py [hunk 1]
```diff
@@ -0,0 +1,1076 @@
...
+            x_mask = torch.unsqueeze(sequence_mask(mel_len, x.size(2)), 1)
+
+        self.pos_encoder = PositionalEncoding(
+            d_model=x.size(1), dropout_rate=0, max_len=x.size(2)
+        )
+
+        x = self.pos_encoder(x.transpose(1, 2)).transpose(1, 2)
+        # decoder_input_pitch = self.pos_encoder(
+        #     decoder_input_pitch.transpose(1, 2)
+        # ).transpose(1, 2)
```
Comment (ftshijt) [inline espnet2/gan_svs/pits/pisinger_generator.py:626]:
```suggestion
```

**Review by ftshijt — COMMENTED** [review]
Many thanks for your contribution!

The general design looks good to me. Given that many components are from TTS papers, probably after we make it stable, the components can be also added back to TTS as well.

**Comment by ftshijt** [comment]
Also given that you have introduced several new modules (great job!), could you please also add more module descriptions as comments in the codebase? As some of the names are not as obvious.

**Comment by jerryuhoo** [comment]
> Many thanks for your contribution!
> 
> The general design looks good to me. Given that many components are from TTS papers, probably after we make it stable, the components can be also added back to TTS as well.

Sure, I am considering removing PISinger code since it doesn't perform as well as VISinger2. And in this PR, I added modules for preprocessing pits feature (yingram), but I will not implement PITS (TTS) in this PR. I plan to add PITS (TTS) in the next PR. Then I will move some common modules for both TTS and SVS such as avocodo into gan_tts folder.

File: egs2/opencpop/svs1/conf/tuning/train_vits.yaml [hunk 2]
```diff
@@ -13,12 +13,13 @@
 svs: vits
```
[the remark this answers is not in the document]
Comment (jerryuhoo) [reply egs2/opencpop/svs1/conf/tuning/train_vits.yaml:13]: Should I also create new yaml files for vocoders such as UHifiGAN and Avocodo?

**Comment by ftshijt** [comment]
> > Many thanks for your contribution!
> > The general design looks good to me. Given that many components are from TTS papers, probably after we make it stable, the components can be also added back to TTS as well.
> 
> Sure, I am considering removing PISinger code since it doesn't perform as well as VISinger2. And in this PR, I added modules for preprocessing PITS feature (yingram), but I will not implement PITS (TTS) in this PR. I plan to add PITS (TTS) in the next PR. Then I will move some common modules for both TTS and SVS such as avocodo into gan_tts folder.

Sounds very cool! Btw, the code for adding slur is merged just now.

**Comment by ftshijt** [comment]
Hi @jerryuhoo could you let me know when you have finished the development? Then I can also help to fix the CI issue in the import test for you.

**Comment by jerryuhoo** [comment]
> Hi @jerryuhoo could you let me know when you have finished the development? Then I can also help to fix the CI issue in the import test for you.

Those listed models and functions are done, but I'm still investigating the performance gap. It is caused by either the posterior encoder or the vocoder, but currently I cannot find the bug.

File: egs2/opencpop/svs1/conf/tuning/train_vits.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (ftshijt) [reply egs2/opencpop/svs1/conf/tuning/train_vits.yaml:13]: Sounds cool!

**Comment by ftshijt** [comment]
Sorry that I did not find time to fix the CI, let's discuss the details in today's meeting

**Comment by jerryuhoo** [comment]
Some code can be improved such as ddsp module (some part of the ddsp code is not used) and modules in MFD. For example, in visinger2_vocoder.py, maybe we can use LogMelFbank instead of TorchSTFT. But LogMelFbank doesn't have a feature of domain="double", which considers both linear and log fbanks.

**Comment by ftshijt** [comment]
> Some code can be improved such as ddsp module (some part of the ddsp code is not used) and modules in MFD. For example, in visinger2_vocoder.py, maybe we can use LogMelFbank instead of TorchSTFT. But LogMelFbank doesn't have a feature of domain="double", which considers both linear and log fbanks.

You may add TODOs to the codebase. As this PR consists of important bug fix, I believe we can quickly merge it by fixing the ci.

**Comment by ftshijt** [comment]
Last request for this PR, can we fix the ci tests for the imported functions https://github.com/espnet/espnet/actions/runs/5043122127/jobs/9044514638 ? After that, I can merge it

File: espnet2/gan_tts/hifigan/hifigan.py [hunk 3]
```diff
@@ -566,7 +566,7 @@ def apply_spectral_norm(self):
         """Apply spectral normalization module from all of the layers."""
 
         def _apply_spectral_norm(m: torch.nn.Module):
-            if isinstance(m, torch.nn.Conv2d):
+            if isinstance(m, torch.nn.Conv1d):
```
Comment (ftshijt) [inline espnet2/gan_tts/hifigan/hifigan.py:569]: https://github.com/espnet/espnet/issues/5215 Hi @jerryuhoo , could you double check the issue?

**Pull request merged**